### DISCUSSION TITLE
'{$nonce}' is not a valid attribute name

### DISCUSSION THREAD

**Discussion by AkramZerarka** [livewire/livewire#2989] [7. Bugs (archived)]
### Description
Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element': '{$nonce}' is not a valid attribute name.

**Context**
- Livewire version: [^2.5.1]
- Laravel version: [^8.47.0]
- Alpine version: [^3.0.7]
- Browser: [Chrome,]

**Comment by orware** [comment]
I'm not sure if I agree with using the discussions here since it seems to imply that my reply here is an "answer" to the problem when it isn't, but here goes...

I'm kind of in a similar boat myself at the moment, although reverting both Livewire and Alpine down to 2.4.4 and the current v2 version of Alpine didn't seem to fix the issue, at least from me quickly testing things, performing some hard refreshes and seeing the same issue occur so I can't be sure if it's tied to the new Livewire update or if it's Alpine.

In my case, it seems to be related to a wire:click call that was previously working:
```
wire:click="setTerm({{ $headingTitleAttribute }})"
```

But when I click on that element it throws that same error:
```
Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element': '@click' is not a valid attribute name
```

And the Chrome DevTools mention morphAttrs.js Line 42 as being the spot it runs into the issue.

**Reply by orware** [reply]
I have the same `wire:click` usage on a different Livewire component and it seems to partially work in comparison to the above situation.

Instead of receiving the error above, I see:
```
Uncaught TypeError: Cannot read property 'startsWith' of undefined
```

And that's triggered in cdn.min.js for Alpine v3 on line 5 (not that it's that helpful since it's a minified file).

In the previous example, the click isn't triggered correctly due to the error, but for this example, the click seems to work (it updates the query string with the new term value and loads up the results that are supposed to be displayed), but it doesn't work as it had been (this morning I was switching some usage of the Bootstrap Collapse feature to possibly utilize Alpine instead so on this subsequent page that loads, all of the Alpine functionality isn't behaving as expected.

However, if I refresh my page at this point, then upon reload everything is working fine so it doesn't actually seem to be an output issue but there does seem to be some sort of interaction error.

For this second situation I can probably dig a little deeper and maybe figure out a solution for it, but for the first situation (the `Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element'` issue) I'm not sure if I'll be able to figure out a solution for that one.

**Reply by orware** [reply]
I still haven't come up with a solution for the `Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element': '@click' is not a valid attribute name` issue, but for the `Uncaught TypeError: Cannot read property 'startsWith' of undefined` issue I was able to switch out from the minified version and learn that the problem was occurring around line 488.

@calebporzio, this would appear to be related to these lines in the AlpineJS source (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L93-L97).

On my end I adjusted things to add in the extra undefined check below which seemed to allow me to bypass that issue (I had added some logging as well, but it generates quite a bit of output, and even with that I wasn't clear on where exactly it was failing, so I'm hoping adding the extra undefined type check is an adequate fix for this particular problem).
```
var startingWith = (subject, replacement) => ({name, value}) => {

      if (typeof name !== 'undefined') {
        if (name.startsWith(subject))
            name = name.replace(subject, replacement);
      }

    return {name, value};
  };
```

I was hoping that this fix might fix the first issue as well, but no luck there yet.

I have gotten a wee bit closer though since I've figured out which part of my Livewire Component is generating the problem...it seems to be related to two different Blade components I'm loading prior to the sections using `wire:click`, but it's really not clear as to why those are causing any problems right now since they are relatively simple Blade components, and I even stripped them down significantly and it still triggered the error so I'm not sure what's up.

Here is one example.

The Livewire component loads the following Blade component in:
```
<x-schedule.new-sections :data="$newSections" :show="$showNewSectionsArea" />
```

And the Blade component itself looks like this (a simple Bootstrap dismissable alert):
```
@props([
    'data',
    'show' => true
])

@if ($data->isNotEmpty() && $show)
    <div class="alert alert-warning alert-dismissible fade show" role="alert">
        <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
        <p class="mb-0"><strong>Check out the following new courses:</strong></p>
        <ul class="mb-0">
            <?php foreach($data as $newSection) : ?>
                <li>
                    <a href="<?php echo '?term=' . $newSection->term_code . '&amp;crn=' . $newSection->crn; ?>"><?php echo  $newSection->subject . ' ' .  $newSection->course_number . ' (CRN: ' . $newSection->crn . ')'; ?></a>
                    -
                    Taught by <?php echo ((!empty($newSection->last_name) && !empty($newSection->first_name)) ? '<a href="' . '?instructor=' . $newSection->username . '">' . $newSection->first_name . ' ' . $newSection->last_name . '</a>' : 'Staff'); ?>
                </li>
            <?php endforeach; ?>
        </ul>
    </div>
@endif
```

But even if I reduce the Blade component to just a simple div:
```
@props([
    'data',
    'show' => true
])

@if ($data->isNotEmpty() && $show)
    <div>Hi!</div>
@endif
```

I still am receiving the `Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element': '@click' is not a valid attribute name.` error on morphAttrs.js line 42.

It's only when I completely remove the call to load the Blade component from the Livewire component that it's able to behave normally as it had been.

**Reply by orware** [reply]
@AkramZerarka and @calebporzio ,

This one might be related to this older item here:
https://github.com/livewire/livewire/issues/828

I don't think an actual solution was noted in that issue item, the only thing that was shared was that usage of `@click` should be changed to `x-on:click` to workaround the issue.

That appears to work for me to get past the `Uncaught (in promise) DOMException: Failed to execute 'setAttribute' on 'Element': '@click' is not a valid attribute name.` issue.

In order to verify if that might also fix the second issue I had ran into, I gave it a shot (reverting back to the latest AlpineJS CDN URL without the undefined type check fix I had added into my local copy) and the `Uncaught TypeError: Cannot read property 'startsWith' of undefined` is still present after switching my output to use `x-on:click` so that might still be a separate bugfix that requires some attention in AlpineJS.

@AkramZerarka, hopefully you are using `@click` or some other option that starts with `@` and this will be a solution for your problem you initially reported above (although now that I'm thinking about it, it's possible your $nonce issue is more PHP related and the variable name is being output into the HTML rather than the variable value).

**Reply by calebporzio** [reply]
I don't think the original bug being posted is related to the `@` symbol error, however, that is a known bug that needs to be fixed. I just merged in a fix: https://github.com/livewire/livewire/pull/3152

Please pull down `dev-master` and verify that that issue is fixed for you. Thanks!
